### PR TITLE
Pv2 1382 apprenticeship price episode date to match stop date

### PR DESCRIPTION
**Pull request by samuel-dev-dba**

### CODE REVIEW

File: src/SFA.DAS.Payments.DataLocks.Domain/Services/CourseValidation/CourseValidationProcessor.cs [hunk 1]
```diff
@@ -5,60 +5,70 @@
...
-            var dataLockFailures = new List<DataLockFailure>();
-            var invalidApprenticeshipPriceEpisodeIds = new List<long>();
-
-            var allApprenticeshipPriceEpisodeIds = dataLockValidationModel.Apprenticeship.ApprenticeshipPriceEpisodes
-                .Select(x => x.Id).ToList();
+            var allApprenticeshipPriceEpisodeIds = GetAllApprenticeshipPriceEpisodeIds(dataLockValidationModel);
 
-
-            foreach (var courseValidator in courseValidators)
+            var startDateValidationResult = Validate(startDateValidator, dataLockValidationModel, allApprenticeshipPriceEpisodeIds);
```
Comment (kuzinator) [inline src/SFA.DAS.Payments.DataLocks.Domain/Services/CourseValidation/CourseValidationProcessor.cs]: This looks like a repetative task. Is this the same as to call `Validate` with the list of validators where the first three validators are StartDtate, Stopped and IncentiveStopped?

File: src/SFA.DAS.Payments.DataLocks.Domain/Services/CourseValidation/EarningPeriodsValidationProcessor.cs [hunk 2]
```diff
@@ -135,5 +156,34 @@ private EarningPeriod CreateEarningPeriod(EarningPeriod period)
...
+            if (apprenticeship.StopDate.HasValue && (apprenticeship.StopDate.Value < periodDates.periodStartDate))
+            {
+                return false;
+            }
+            
+            var latestApprenticeshipPriceEpisode = apprenticeship
+                .ApprenticeshipPriceEpisodes
+                .Where(x => !x.Removed)
+                .OrderByDescending(x => x.EndDate)
+                .First();
```
Comment (kuzinator) [inline src/SFA.DAS.Payments.DataLocks.Domain/Services/CourseValidation/EarningPeriodsValidationProcessor.cs]: You might be processing an earning from a few months back, getting the latest apprenticeship price episode will be wrong thing to do.

File: src/SFA.DAS.Payments.DataLocks.Domain/Services/CourseValidation/EarningPeriodsValidationProcessor.cs [hunk 3]
```diff
@@ -64,20 +68,37 @@ public EarningPeriodsValidationProcessor(ICourseValidationProcessor courseValida
...
                 }
+                
+                var apprenticeshipsWithinPeriod = apprenticeships
+                    .Where(a => a.Ukprn == ukprn && ApprenticeshipIsWithinPeriod(a, period.Period, academicYear))
+                    .ToList();
 
-                foreach (var apprenticeship in apprenticeships.Where(apprenticeship => apprenticeship.Ukprn == ukprn))
+                if (!apprenticeshipsWithinPeriod.Any())
+                {
+                    period.DataLockFailures = new List<DataLockFailure>
```
[the remark this answers is not in the document]
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.DataLocks.Domain/Services/CourseValidation/EarningPeriodsValidationProcessor.cs]: Yes we are handling 0 amount  in  line 66

File: src/SFA.DAS.Payments.DataLocks.Domain/Services/CourseValidation/EarningPeriodsValidationProcessor.cs [hunk 4]
```diff
@@ -135,5 +156,34 @@ private EarningPeriod CreateEarningPeriod(EarningPeriod period)
...
+
+            if (apprenticeship.StopDate.HasValue && (apprenticeship.StopDate.Value < periodDates.periodStartDate))
+            {
+                return false;
+            }
+            
+            var latestApprenticeshipPriceEpisode = apprenticeship
+                .ApprenticeshipPriceEpisodes
+                .Where(x => !x.Removed)
+                .OrderByDescending(x => x.EndDate)
```
[the remark this answers is not in the document]
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.DataLocks.Domain/Services/CourseValidation/EarningPeriodsValidationProcessor.cs]: Good point Changed

File: src/SFA.DAS.Payments.DataLocks.Domain/Services/CourseValidation/CourseValidationProcessor.cs (hunk 1 above)
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.DataLocks.Domain/Services/CourseValidation/CourseValidationProcessor.cs]: Good point now changed

File: src/SFA.DAS.Payments.DataLocks.Domain/Services/CourseValidation/EarningPeriodsValidationProcessor.cs (hunk 2 above)
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.DataLocks.Domain/Services/CourseValidation/EarningPeriodsValidationProcessor.cs]: Changed

**Pull request merged**